src/subprocess-posix.cc

```diff
@@ -169,11 +170,61 @@ void SubprocessSet::HandlePendingInterruption() {
     interrupted_ = SIGTERM;
 }
 
+bool SubprocessSet::child_exited_;
+
+void SubprocessSet::SetChildExited(int signum) {
+  child_exited_ = true;
+}
+
+void SubprocessSet::HandleChildExit() {
+  if (!child_exited_)
+    return;
+  child_exited_ = false;
+  for (int i = running_.size(); i > 0; --i)
+    WaitChildExit(false);
+}
+
+void SubprocessSet::WaitChildExit(bool block) {
+  int status = 0;
+  int options = block ? 0 : WNOHANG;
+  pid_t pid = waitpid(-1, &status, options);
+  if (pid == -1)
+    Fatal("waitpid: %s", strerror(errno));
+  if (pid == 0)
+    return;
+  for (vector<Subprocess*>::iterator i = running_.begin();
+       i != running_.end(); ++i) {
+    Subprocess *subprocess = *i;
+    if (pid != subprocess->pid_)
+      continue;
+    // Drain pipe by reading until we would block, as
+    // finished processes' pipes are not read by DoWork.
+    while (subprocess->fd_ != -1) {
+      pollfd pfd = { subprocess->fd_, POLLIN | POLLPRI, 0 };
+      int n = poll(&pfd, 1, 0);
```

Comment:
.. if this draining loop can be removed, that would void the need to add #ifdef USE_PPOLL / USE_SELECT here for consistency with the rest of the code. Otherwise, we probably need to #ifdef to use either select or poll here.